lib/fact_changes.rb

```diff
@@ -535,38 +522,31 @@ def _fact_operations(action_type, step, facts)
     operations = facts.map do |fact|
       modified_assets.push(fact.object_asset) if listening_to_predicate?(fact.predicate)
       Operation.new(:action_type => action_type, :step => step,
-        :asset=> fact.asset, :predicate => fact.predicate, :object => fact.object, object_asset: fact.object_asset)
+                    :asset => fact.asset, :predicate => fact.predicate, :object => fact.object, object_asset: fact.object_asset)
     end
 
     modified_assets.flatten.compact.uniq.each(&:touch)
     operations
   end
 
-  def all_values_are_new_records(hash)
-    hash.values.all? do |value|
-      (value.respond_to?(:new_record?) && value.new_record?)
-    end
+  def any_values_are_new_records(hash)
+    hash.values.any? { |value| value.try(:new_record?) }
   end
 
   def _instance_builder_for_import(klass, params_list, &block)
-
-    instances = params_list.map do |params_for_instance|
-      unless (params_for_instance.kind_of?(klass))
-        if (all_values_are_new_records(params_for_instance) ||
-          (!klass.exists?(params_for_instance)))
-          klass.new(params_for_instance)
-        end
-      else
-        if params_for_instance.new_record?
-          params_for_instance
-        end
+    instances = params_list.filter_map do |params_for_instance|
+      if params_for_instance.kind_of?(klass)
+        params_for_instance if params_for_instance.new_record?
+      elsif any_values_are_new_records(params_for_instance) || !klass.exists?(params_for_instance)
```

Comment:
`klass.exists?(params_for_instance)` is problematic, and is causing a *lot* of extra database hits. I've left it for now, as this code is quite low level, so hard to make assumptions about (klass could be many things for instance).